:seedling: add fuzzing targets for IPAM

as20203 opened the pull request:
This PR add fuzzing targets for IPAM repository which will later be used by OSS-Fuzz later on.

tuminoid commented:
Just using this PR to trigger 1.4 test once.

/test-ubuntu-integration-release-1-4

kashifest commented:
/retest

kashifest commented:
/cc
@adilGhaffarDev @Rozzii @lentzi90 @mboukhalfa @smoshiur1237 @Sunnatillo @mquhuy @kashifest

as20203 commented:
/hold

adilGhaffarDev commented:
Few things that should be changed:

- Change name of the functions: `IPPoolFinalizer` to `FuzzNewIPPoolManager` (since we are fuzzing `NewIPPoolManager`) and `IPClaimToPool` to `FuzzIPClaimToIPPool`
- Filenaming: Tests are from two different files, we should put them is separate files. Place `FuzzNewIPPoolManager` in `ippool_manager_fuzzer.go` and FuzzIPClaimToIPPool in `ippool_controller_fuzzer.go`. 

I would also suggest to check how CAPI is implmenting fuzz tests and try to replicate same. Kindly check this: https://github.com/cncf/cncf-fuzzing/blob/main/projects/cluster-api/cluster_controller_fuzzer.go

as20203 commented:
Updated the file and function naming as well as updated the test target similar to [](https://github.com/cncf/cncf-fuzzing/blob/main/projects/cluster-api/cluster_controller_fuzzer.go) as mentioned by @adilGhaffarDev .

as20203 commented:
/rerun

tuminoid commented:
/retest

lentzi90 reviewed (commented):
Looks promising!
How do we run this? I think we will need some kind of minimal readme at least to explain how to use it. I guess we also want to make it part of the unit tests we usually run?

as20203 commented:
/golangci-lint

tuminoid commented:
> Looks promising! How do we run this? I think we will need some kind of minimal readme at least to explain how to use it. I guess we also want to make it part of the unit tests we usually run?

Fuzzing targets aren't unit test targets, they often take hours to run. The plan is to add project to OSS-Fuzz https://google.github.io/oss-fuzz/getting-started/accepting-new-projects/ but they have acceptance criteria we may not meet. In that case, we need to add a fuzzing Jenkins job.

as20203 commented:
I'll add a minimal Readme file in the fuzz repository for this PR explaining how to run these fuzz targets. I've also mentioned them as a comment on the ticket as well.

lentzi90 commented:
> Fuzzing targets aren't unit test targets, they often take hours to run. The plan is to add project to OSS-Fuzz https://google.github.io/oss-fuzz/getting-started/accepting-new-projects/ but they have acceptance criteria we may not meet. In that case, we need to add a fuzzing Jenkins job.

I see, thanks for explaining! My confusion comes from the kind of fuzzing we do in [CAPO](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/main/api/v1alpha6/conversion_test.go) and [CAPM3](https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/api/v1alpha5/conversion_test.go) for API version conversion. These are part of the unit tests.

mquhuy commented:
Could you squash all the fixing commits onto the original one and force push? It would be much easier to review that way.

as20203 commented:
I've added a readme inside the fuzz directory for running these test functions locally. Other than that, I've also squashed all the fixing commits with the original one as mentioned by @mquhuy

mquhuy commented:
/test-centos-e2e-integration-main
/test-ubuntu-integration-main

lentzi90 reviewed (commented):
/approve

kashifest commented on test/fuzz/README.md:
```diff
@@ -0,0 +1,30 @@
+# Fuzzing
+
+So I made a fuzzing target and run a fuzzing check for a function in ip_address_manager for testing finalizers.  Following are the important steps in this regard:
```
Lets generalize this first paragraph a bit, perhaps not mentioning `I` and say one or two lines about Fuzzing in general.

as20203 replied:
sure @kashifest , I'll update this .

as20203 replied:
The description is updated now for README.

mquhuy commented:
/test-centos-e2e-integration-main
/test-ubuntu-integration-main

kashifest commented:
/lgtm

tuminoid commented on go.mod:
```diff
@@ -21,7 +21,10 @@ require (
 
 )
 
+require github.com/cyphar/filepath-securejoin v0.2.3 // indirect
```
Why is this separately from other indirects?

tuminoid commented on test/fuzz/README.md:
```diff
@@ -0,0 +1,30 @@
+# Fuzzing
+
+ Fuzzing or fuzz testing is an automated software testing technique that involves providing invalid, unexpected, or random data as inputs to a computer program. A fuzzing target is defined for running the tests. These targets are defined in the test/fuzz directory and run using the instructions given below:
```
Change the line length to 80 (why isn't markdownlint catching this...?).

lentzi90 replied:
Could it be that we don't check files in nested folders? :scream:

as20203 replied:
running markdown lint on this file isn't giving any errors for line length. I tried adding new line breaks to change length but then it is giving trailing space issues here. When I tried adding br tags which is another way it was also giving linting error. What should be the way to resolve this since I don't receive any error for this rule.

tuminoid replied:
MDL had obsolete config file, which ignored line length. It'll be fixed here: https://github.com/metal3-io/ip-address-manager/pull/252

tuminoid replied:
OK, it has merged. Please rebase and also squash commits.

tuminoid reviewed (commented):
Nits.

as20203 commented on go.mod:
```diff
@@ -21,7 +21,10 @@ require (
+require github.com/cyphar/filepath-securejoin v0.2.3 // indirect
+
 require (
+	github.com/AdaLogics/go-fuzz-headers v0.0.0-20230106234847-43070de90fa1
```
I've fixed this dependency location and will push this change.

lentzi90 commented on test/fuzz/README.md:
```diff
@@ -0,0 +1,53 @@
+    Unmarshal and then run the function. It keeps on running
+    until it finds a bug.
+
+    `[63 32 45 58 119 50 9 119 50 9 119 50 9 45 58
+    50 9 119 9 119 50 9 91 13 45 58]
+    &{TypeMeta:{Kind: APIVersion:}
+    ObjectMeta:{Name: GenerateName: Namespace: SelfLink: UID:
+    ResourceVersion: Generation:0
+    CreationTimestamp:0001-01-01 00:00:00 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil>
+    Labels:map[] Annotations:map[] OwnerReferences:[]
+    Finalizers:[] ManagedFields:[]} Spec:{ClusterName:<nil> Pools:[]
+    PreAllocations:map[] Prefix:0 Gateway:<nil> DNSServers:[]
+    NamePrefix:} Status:{LastUpdated:<nil> Allocations:map[]}}
+    #1028919 REDUCE cov: 4365 ft: 18232 corp:
+    2348/250Kb lim: 1070 exec/s: 3535 rss: 463Mb L:
+    118/1060 MS: 1 EraseBytes-
+`
```
This looks a bit weird. I would suggest putting it in a proper code block with triple backticks (```) and same for b and c above.

as20203 replied:
I've updated this as per your suggestion.

lentzi90 commented on test/fuzz/README.md:
````diff
@@ -0,0 +1,54 @@
+    #1028919 REDUCE cov: 4365 ft: 18232 corp:
+    2348/250Kb lim: 1070 exec/s: 3535 rss: 463Mb L:
+    118/1060 MS: 1 EraseBytes-
+```
````
```suggestion
    ```
```

as20203 replied:
fixed this change.

lentzi90 commented on test/fuzz/README.md:
```diff
@@ -0,0 +1,54 @@
+
+    b. Run below command to make C binary files for the same function.
+
+    `clang -o yaml_FuzzNewIPPoolManager  yaml_FuzzNewIPPoolManager.a -fsanitize=fuzzer
+    `
```
This should also be in triple backticks.

as20203 replied:
resolved.

lentzi90 commented on test/fuzz/README.md:
```diff
@@ -0,0 +1,54 @@
+
+    c. Run the fuzzer as below
+
+    `./yaml_FuzzNewIPPoolManager
+`
```
And this

lentzi90 commented on test/fuzz/README.md:
````diff
@@ -0,0 +1,57 @@
+    ```
+    ./yaml_FuzzNewIPPoolManager
+    ```
+`
````
Seems like markdownlint is upset about this left over :smile: 
```suggestion
```

lentzi90 commented:
Looks good from my point :slightly_smiling_face:

mquhuy commented on test/fuzz/README.md:
````diff
@@ -0,0 +1,56 @@
+    ```
+
+- Execution: The execution of the fuzzer for my sample code kept on
+  running with below sample outputs:
````
Avoid using subjective phrases ("my sample code"). Instead, something like "Here is an example of test outputs:"
Move the explanation part to below the output btw.

as20203 replied:
resolved this issue.

tuminoid commented:
/test-ubuntu-integration-main  
/test-centos-e2e-integration-main

tuminoid commented:
/lgtm 

Let's get this merged, and expand later. Also need to add Jenkins jobs and/or apply for the OSS-Fuzz inclusion.

as20203 commented:
/unhold
